src/Castle.Core/DynamicProxy/IInterceptorBase.cs

```diff
@@ -0,0 +1,56 @@
+﻿
+namespace Castle.DynamicProxy
+{
+	using System;
+
+	public class ExceptionContext
+	{
+
+		public Exception Ex { get; set; }
+		public IInvocation Invocation { get; set; }
+	}
+
+	public abstract class InterceptorBase : IInterceptor
+	{
+		protected virtual void OnExcuting(IInvocation invocation)
+		{
+		}
+
+		protected virtual void OnExcuted(IInvocation invocation)
+		{
+		}
+
+		protected virtual void OnException(ExceptionContext context)
+		{
+		}
+
+		private bool NotVoidHasValue(IInvocation invocation)
+		{
+			return !(invocation.Method.ReturnType == typeof(void) ||
+			         invocation.ReturnValue == null);
+		}
+
+		public void Intercept(IInvocation invocation)
+		{
+			try
+			{
+				OnExcuting(invocation);
+
+				if (NotVoidHasValue(invocation))
+					return;
+
+				invocation.Proceed();
```

Comment:
I don't think that this pattern&mdash;`Proceed` if the interceptor hasn't yet provided a return value&mdash;holds true for all use cases (though admittedly for a good many). Sometimes there might be nothing to proceed *to*. You'd at least have to document your interceptor's behavior / functional semantics.